ocpmodels/common/utils.py

```diff
@@ -355,6 +356,186 @@ def get_pbc_distances(
+    pos1 = torch.index_select(atom_pos, 0, index1)
+    pos2 = torch.index_select(atom_pos, 0, index2)
+
+    # Tensor of unit cells. Assumes 9 cells in -1, 0, 1 offsets in the x and y dimensions
+    unit_cell = torch.tensor(
+        [
+            [-1, -1, 0],
+            [-1, 0, 0],
+            [-1, 1, 0],
+            [0, -1, 0],
+            [0, 0, 0],
+            [0, 1, 0],
+            [1, -1, 0],
+            [1, 0, 0],
+            [1, 1, 0],
+        ],
+        device=device,
+    ).float()
+    num_cells = len(unit_cell)
+    unit_cell_per_atom = unit_cell.view(1, num_cells, 3).repeat(
+        len(index2), 1, 1
+    )
+    unit_cell = torch.transpose(unit_cell, 0, 1)
+    unit_cell_batch = unit_cell.view(1, 3, num_cells).expand(
+        batch_size, -1, -1
+    )
+
+    # Compute the x, y, z positional offsets for each cell in each image
+    data_cell = torch.transpose(data.cell, 1, 2)
+    pbc_offsets = torch.bmm(data_cell, unit_cell_batch)
+    pbc_offsets_per_atom = torch.repeat_interleave(
+        pbc_offsets, num_atoms_per_image_sqr, dim=0
+    )
+
+    # Expand the positions and indices for the 9 cells
+    pos1 = pos1.view(-1, 3, 1).expand(-1, -1, num_cells)
+    pos2 = pos2.view(-1, 3, 1).expand(-1, -1, num_cells)
+    index1 = index1.view(-1, 1).repeat(1, num_cells).view(-1)
+    index2 = index2.view(-1, 1).repeat(1, num_cells).view(-1)
+    # Add the PBC offsets for the second atom
+    pos2 = pos2 + pbc_offsets_per_atom
+
+    # Compute the squared distance between atoms
+    atom_distance_sqr = torch.sum((pos1 - pos2) ** 2, dim=1)
+    atom_distance_sqr = atom_distance_sqr.view(-1)
+
+    # Remove pairs that are too far apart
+    mask_within_radius = torch.le(atom_distance_sqr, radius * radius)
+    # Remove pairs with the same atoms (distance = 0.0)
+    mask_not_same = torch.gt(atom_distance_sqr, 0.0001)
+    mask = torch.logical_and(mask_within_radius, mask_not_same)
+    index1 = torch.masked_select(index1, mask)
+    index2 = torch.masked_select(index2, mask)
+    unit_cell = torch.masked_select(
+        unit_cell_per_atom.view(-1, 3), mask.view(-1, 1).expand(-1, 3)
+    )
+    unit_cell = unit_cell.view(-1, 3)
+
+    num_atoms = len(data.pos)
+    num_neighbors = torch.zeros(num_atoms, device=device)
+    num_neighbors.index_add_(0, index1, torch.ones(len(index1), device=device))
+    num_neighbors = num_neighbors.long()
+    max_num_neighbors = torch.max(num_neighbors).long()
+
+    # Compute neighbors per image
+    _max_neighbors = copy.deepcopy(num_neighbors)
+    _max_neighbors[_max_neighbors > max_num_neighbors_threshold] = max_num_neighbors_threshold
+    _num_neighbors = torch.zeros(num_atoms + 1, device=device).long()
+    _natoms = torch.zeros(data.natoms.shape[0] + 1, device=device).long()
+    _num_neighbors[1:] = torch.cumsum(_max_neighbors, dim=0)
+    _natoms[1:] = torch.cumsum(data.natoms, dim=0)
+    num_neighbors_image = _num_neighbors[_natoms[1:]] - _num_neighbors[_natoms[:-1]]
```

Comment:
@anuroopsriram The neighbors_image implementation you had hit errors for me sometimes, let me know if this makes sense.